In [1]:
print('นางสาววิภาวดี แก้วดวงดี')
print('รหัส 603021112-8')

นางสาววิภาวดี แก้วดวงดี
รหัส 603021112-8


ข้อ 1 สร้าง function ชื่อ smartDict ซึ่งทำหน้าที่ :
     function รับ input เป็น list 2ตัวที่มีขนาดเท่ากันโดย list แรกเป็น lisของตัวเลข list ที่สอง เป็นlist ของ text
    ถ้าค่าใน list แรกเป็นเลขคี่ ให้ค่าใน dict ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 1และ value จาก list ตัวที่ 2
     ถ้าค่าใน list แรกเป็นเลขคู่ ให้ค่าใน dict ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 2และ value จาก list ตัวที่ 1
    output ของ function คือ dict ที่สร้างขึ้น
    เช่น
    output = smartDict ([1,2,3],['apple','banana','orange'])
    จะได้ outout1 เป็น {1:'apple','banana':2,3:'orange'}

In [263]:
def smartDict(text,num,):
    f = {}
    for i in range(len(num)):
        if  i == True:
            f[num[i]] = text[i]
        else:
            f[text[i]] = num[i]
    return(f)        

In [264]:
smartDict ([1,2,3],['apple','banana','orange'])

{1: 'apple', 'banana': 2, 3: 'orange'}

In [265]:
smartDict ([7,8,9],['a','b','o'])

{7: 'a', 'b': 8, 9: 'o'}

จากไฟล์ Wuhan_nCov_GNome.tst จงเขียนโปรแกรมเพื่อตอบคำถามต่อไปนี้
1กรดอะมิโนใดใน (A,T,G,C) มีปริมาณมากที่สุดและมีเท่าใด
2จงหา sequence ที่มีขนาด 3 และ 4 ที่เกิดขึ้นซ้ำมากที่สุด


In [4]:
import pandas as pd

In [5]:
DNAseq = pd.read_csv('C:\\Python\\Wuhan_nCoV_GNome.txt',header=None) 
DNAseq.head()

0
0  ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGA...
1  CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTG...
2  TAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGC...
3  TTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCG...
4  CCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTT...

## กรดอะมิโนใดใน (A,T,G,C) มีปริมาณมากที่สุดและมีเท่าใด

In [6]:
DNAseq.shape

(428, 1)

In [7]:
def countChar(seq,character):
    charcount = 0 
    for char in seq:
        if char == character: 
            charcount = charcount + 1 
    return charcount 

In [8]:
ANum = 0 
TNum = 0
GNum = 0
CNum = 0
for row in range(DNAseq.shape[0]): 
    ANum += countChar(DNAseq.iloc[row][0],'A') 
    TNum += countChar(DNAseq.iloc[row][0],'T')
    GNum += countChar(DNAseq.iloc[row][0],'G')
    CNum += countChar(DNAseq.iloc[row][0],'C')
print(f'มีกรดอมิโน A T G C เท่ากับ  {ANum} As, {TNum} Ts, {GNum} Gs, {CNum} Cs ตามลำดับ')

มีกรดอมิโน A T G C เท่ากับ  8954 As, 9594 Ts, 5863 Gs, 5492 Cs ตามลำดับ


#### Ans1 กรดอะมิโน  T  มีมากที่สุด มี 9594 ตัว

## จงหา sequence ที่มีขนาด 3 และ 4 ที่เกิดขึ้นซ้ำมากที่สุด

### สามตัวติดกัน

In [13]:
def pop3(seq): 
    popula3 = list() 
    for i in range(len(seq)-2): 
        popula3.append(seq[i:i+3]) 
    return popula3

In [14]:
pop3('absaabsabbs')

['abs', 'bsa', 'saa', 'aab', 'abs', 'bsa', 'sab', 'abb', 'bbs']

In [16]:
pop3(DNAseq.iloc[5][0])

['GTG',
 'TGG',
 'GGC',
 'GCT',
 'CTT',
 'TTT',
 'TTG',
 'TGG',
 'GGA',
 'GAG',
 'AGA',
 'GAC',
 'ACT',
 'CTC',
 'TCC',
 'CCG',
 'CGT',
 'GTG',
 'TGG',
 'GGA',
 'GAG',
 'AGG',
 'GGA',
 'GAG',
 'AGG',
 'GGT',
 'GTC',
 'TCT',
 'CTT',
 'TTA',
 'TAT',
 'ATC',
 'TCA',
 'CAG',
 'AGA',
 'GAG',
 'AGG',
 'GGC',
 'GCA',
 'CAC',
 'ACG',
 'CGT',
 'GTC',
 'TCA',
 'CAA',
 'AAC',
 'ACA',
 'CAT',
 'ATC',
 'TCT',
 'CTT',
 'TTA',
 'TAA',
 'AAA',
 'AAG',
 'AGA',
 'GAT',
 'ATG',
 'TGG',
 'GGC',
 'GCA',
 'CAC',
 'ACT',
 'CTT',
 'TTG',
 'TGT',
 'GTG',
 'TGG']

In [20]:
list_all3= list()
for row in range(DNAseq.shape[0]): 
    list_all3 = list_all3 + pop3(DNAseq.iloc[row][0]) 

In [22]:
len(list_all3)

29047

In [23]:
unique_3char = list(set(list_all3))

In [24]:
len(unique_3char)

64

In [25]:
dict_3char = {}
for sub3 in unique_3char: 
    num = 0 
    for i in list_all3:
        if sub3 == i: 
            num += 1
        dict_3char[sub3] = num

In [26]:
dict_3char

{'GTT': 686,
 'ACG': 157,
 'ACA': 789,
 'GAG': 290,
 'GGG': 131,
 'AAT': 735,
 'GCT': 509,
 'CAT': 472,
 'ATG': 701,
 'GCC': 184,
 'CTT': 720,
 'AAC': 595,
 'CCA': 336,
 'AGC': 292,
 'CTA': 543,
 'TAT': 601,
 'ATT': 749,
 'GGT': 442,
 'TCG': 110,
 'GCA': 363,
 'TCT': 526,
 'TCA': 536,
 'TAC': 591,
 'GAT': 430,
 'TGC': 530,
 'AGT': 493,
 'GTG': 532,
 'TTG': 800,
 'TAG': 414,
 'TTA': 853,
 'AAA': 899,
 'CAC': 447,
 'GAA': 524,
 'AAG': 564,
 'CCC': 113,
 'CGA': 91,
 'CAA': 681,
 'GCG': 86,
 'TGT': 834,
 'TTC': 509,
 'CGG': 74,
 'CAG': 424,
 'GTA': 458,
 'ACC': 364,
 'TGA': 602,
 'GAC': 328,
 'AGG': 317,
 'CCG': 70,
 'CTC': 276,
 'AGA': 582,
 'CGC': 92,
 'CCT': 332,
 'ACT': 660,
 'GGA': 272,
 'TCC': 203,
 'GGC': 217,
 'GTC': 265,
 'CTG': 483,
 'CGT': 166,
 'TAA': 688,
 'TTT': 984,
 'ATC': 336,
 'TGG': 540,
 'ATA': 456}

### อะมิโน 4 ตัว ติดกัน

In [28]:
def pop4(seq): 
    popula4 = list() 
    for i in range(len(seq)-3): 
        popula4.append(seq[i:i+4]) 
    return popula4

In [29]:
pop4('abcabbcabbscab')

['abca',
 'bcab',
 'cabb',
 'abbc',
 'bbca',
 'bcab',
 'cabb',
 'abbs',
 'bbsc',
 'bsca',
 'scab']

In [34]:
pop4(DNAseq.iloc[5][0])

['GTGG',
 'TGGC',
 'GGCT',
 'GCTT',
 'CTTT',
 'TTTG',
 'TTGG',
 'TGGA',
 'GGAG',
 'GAGA',
 'AGAC',
 'GACT',
 'ACTC',
 'CTCC',
 'TCCG',
 'CCGT',
 'CGTG',
 'GTGG',
 'TGGA',
 'GGAG',
 'GAGG',
 'AGGA',
 'GGAG',
 'GAGG',
 'AGGT',
 'GGTC',
 'GTCT',
 'TCTT',
 'CTTA',
 'TTAT',
 'TATC',
 'ATCA',
 'TCAG',
 'CAGA',
 'AGAG',
 'GAGG',
 'AGGC',
 'GGCA',
 'GCAC',
 'CACG',
 'ACGT',
 'CGTC',
 'GTCA',
 'TCAA',
 'CAAC',
 'AACA',
 'ACAT',
 'CATC',
 'ATCT',
 'TCTT',
 'CTTA',
 'TTAA',
 'TAAA',
 'AAAG',
 'AAGA',
 'AGAT',
 'GATG',
 'ATGG',
 'TGGC',
 'GGCA',
 'GCAC',
 'CACT',
 'ACTT',
 'CTTG',
 'TTGT',
 'TGTG',
 'GTGG']

In [35]:
list_all4= list()
for row in range(DNAseq.shape[0]): 
    list_all4 = list_all4 + pop4(DNAseq.iloc[row][0]) 

In [36]:
len(list_all4)

28619

In [40]:
unique_4 = list(set(list_all4))

In [41]:
len(unique_4)

256

In [42]:
dict_4 = {}
for sub4 in unique_4: 
    num = 0 
    for i in list_all4:
        if sub4 == i: 
            num += 1
        dict_4[sub4] = num

In [43]:
dict_4

{'AACT': 192,
 'AGGC': 59,
 'GCAA': 112,
 'CTTT': 242,
 'ACAG': 158,
 'GTCA': 96,
 'CGTC': 23,
 'ATCA': 135,
 'ATCT': 124,
 'CCAA': 102,
 'ACGA': 33,
 'ACGT': 61,
 'AAAG': 197,
 'CGTA': 39,
 'CGCT': 27,
 'ACAA': 269,
 'CTCG': 32,
 'TACG': 44,
 'ACCG': 28,
 'GTTG': 194,
 'CTAC': 118,
 'TTGC': 163,
 'ACCC': 55,
 'GTAG': 112,
 'ATCC': 50,
 'CTTC': 128,
 'GGGT': 55,
 'TCCT': 76,
 'GTCG': 19,
 'ACAT': 190,
 'TGAA': 188,
 'CCTG': 63,
 'CCGA': 16,
 'TGGG': 54,
 'AATC': 106,
 'TGTT': 318,
 'AGGA': 87,
 'TTAT': 234,
 'CATC': 92,
 'CGAG': 11,
 'GGCC': 30,
 'TTGA': 185,
 'CGGG': 9,
 'GCGA': 19,
 'GATG': 166,
 'CACG': 39,
 'AGCG': 22,
 'CTCT': 89,
 'TAGC': 76,
 'AGTA': 105,
 'CTAT': 161,
 'TTTT': 294,
 'CCCT': 47,
 'ATTT': 220,
 'AGCA': 83,
 'CGCA': 38,
 'AAAC': 182,
 'CGTT': 45,
 'TTAC': 235,
 'CCGT': 26,
 'TACA': 231,
 'GTCC': 43,
 'TATC': 75,
 'GCGG': 14,
 'TTGT': 282,
 'TGCA': 159,
 'TCGG': 22,
 'AGAC': 113,
 'TCGA': 22,
 'GAAC': 91,
 'ACGG': 28,
 'TCCA': 83,
 'GGAT': 49,
 'CAAC': 181,
 'GGGG'

ข้อ 3 จากไฟล์ data.xlsx สร้างตาราง
ใหม่ที่มี 2 columnโดย column แรกชื่อว่า ภาษาอิสาน และมีค่าในcolumn เป็นค่าจาก column ภาษาอิสานเดิม แต่เปลี่ยน คำที่เป็น "/" เป็น ", " 
และ column ที่สองชื่อว่า ภาษาไทย และมีค่าใน column เป็นค่าจาก column ภาษาไทยเดิมแต่เปลี่ยนค่าที่เป็น "/" เป็น "," เช่น
จาก สวัสดี/สุผู้สุคน/เด้อ/จ้า สวัสดี/ทุกๆคน/นะ/คะ
เป็นสวัสดี,สุผู้สุคน,เด้อ,จ้าสวัสดี,ทุกๆคน,นะ,คะ
สร้างตาราง Dictionary โดนตารางประกอบด้วยสอง column คือ ภาษาอิสาน และ ภาษาไทย โดยค่าใน column ภาษาอิสาน คือคำในภาษาอิสาน และ column ภาษาไทย คือคำแปลเป็นภาษากลางของภาษาอิสานใน 
column แรก (ใช้ค่าจากตาราง data.xlsx โดยคำที่อยู่ในลำดับเดียวกันของcolumns ภาษาอิสาน และภาษาไทยคือคำแปลของกันและกันถ้าจำนวนคำของทั้งสอง columnเท่ากัน - ถ้าจำนวนคำระหว่างสอง column 
ไม่เท่ากันให้ข้าม record นั้นไป)


In [154]:
import pandas as pd

In [171]:
L = ['ภาษาอีสาน' , 'ภาษาไทย']
Data = pd.DataFrame.from_records([],columns = L) 
for i in range(32): 
    df = pd.read_excel('data.xlsx', sheet_name = i, usecols = 'B:C', skipfooter = 1) 
    Data = pd.concat([Data,df], axis = 0, ignore_index = True)

In [172]:
Data

ภาษาอีสาน  \
0                            สวัสดี/สุผู้สุคน/เด้อ/จ้า   
1         กิน/ข้าว/ก่อน/เด้อ/จ้า/ ไผ/สิ/ถ่า/เค้า/ดาวน์   
2    เค้า/ดาว/กะ/ขอ/ให้/เค้า/ดาว/อย่าง/มี/ความ/สุข/...   
3    นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...   
4                       ขอ/ให้/เชิญ/แซ่บ/เด้อ/พี่/น้อง   
..                                                 ...   
970  นั่น/แหละ/คุย/กัน/ไป่/อูย/คุย/กัน/ไป่/เป็น/ปี/...   
971  กะ/เล๊ย/แบบ/ชี/วิต/ชวง/นั่น/แบบ/สิ/เอา/จัง/ได๋...   
972  เส็ด/แล่ว/กะ/เล๊ย/อ่า/เพิ่น/เซิน/ปะ/เออ/เซิน/ม...   
973  ซวง/นั่น/ข่อย/กะ/เขว/เนาะ/จ้า/สิ/ไป่/ตอ/หรือ/ส...   
974  ข่อย/กะ/เล๊ย/ตัด/สิน/ใจ/กั๊บ/มู/หว่า/เออ/ข่อย/...   

                                               ภาษาไทย  
0                                สวัสดี/ทุก ๆ คน/นะ/คะ  
1            กิน/ข้าว/ก่อน/นะ/จ้ะ/ใคร/จะ/รอ/เค้า/ดาวน์  
2    เค้า/ดาวน์/ก็/ขอ/ให้/เค้า/ดาน์/อย่าง/มี/ความ/ส...  
3    นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...  
4                          ขอ/ให้/เชิญ/อร่อย/นะ/ทุก/คน  
..                                                 ...  
970  นั้น/แหละ/คุย/กัน/ไป/อูย/คุย/กัน/ไป/เป็น/ปี/อย...  
971  ก็/เลย/แบบ/ชี/วิต/ช่วง/นั้น/แบบ/จะ/เอา/ยัง/ไง/...  
972  เสร็จ/แล้ว/ก็/เลย/อ่า/เขา/เชิญ/ปะ/เออ/เชิญ/มา/...  
973  ช่วง/นั้น/ฉัน/ก็/เคว้ง/อ่ะ/ค่ะ/จะ/ไป/ต่อ/หรือ/...  
974  ฉัน/ก็/เลย/ตัด/สิน/ใจ/กับ/เพื่อน/ว่า/เออ/ฉัน/จ...  

[975 rows x 2 columns]

In [228]:
C = ['ภาษาอีสาน' , 'ภาษาไทย']
New_table = pd.DataFrame.from_records([],columns = C) 

In [229]:
New_table

Empty DataFrame
Columns: [ภาษาอีสาน, ภาษาไทย]
Index: []

In [230]:
list_lao= [] 
list_thai = [] 
for i in range(len(Data)):
    list_lao.append(replacestr(Data.iloc[i]['ภาษาอีสาน']))
    list_thai.append(replacestr(Data.iloc[i]['ภาษาไทย']))

In [231]:
New_table['ภาษาอีสาน'] = list_lao
New_table['ภาษาไทย'] = list_thai

In [232]:
New_table

ภาษาอีสาน  \
0                            สวัสดี,สุผู้สุคน,เด้อ,จ้า   
1         กิน,ข้าว,ก่อน,เด้อ,จ้า, ไผ,สิ,ถ่า,เค้า,ดาวน์   
2    เค้า,ดาว,กะ,ขอ,ให้,เค้า,ดาว,อย่าง,มี,ความ,สุข,...   
3    นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...   
4                       ขอ,ให้,เชิญ,แซ่บ,เด้อ,พี่,น้อง   
..                                                 ...   
970  นั่น,แหละ,คุย,กัน,ไป่,อูย,คุย,กัน,ไป่,เป็น,ปี,...   
971  กะ,เล๊ย,แบบ,ชี,วิต,ชวง,นั่น,แบบ,สิ,เอา,จัง,ได๋...   
972  เส็ด,แล่ว,กะ,เล๊ย,อ่า,เพิ่น,เซิน,ปะ,เออ,เซิน,ม...   
973  ซวง,นั่น,ข่อย,กะ,เขว,เนาะ,จ้า,สิ,ไป่,ตอ,หรือ,ส...   
974  ข่อย,กะ,เล๊ย,ตัด,สิน,ใจ,กั๊บ,มู,หว่า,เออ,ข่อย,...   

                                               ภาษาไทย  
0                                สวัสดี,ทุก ๆ คน,นะ,คะ  
1            กิน,ข้าว,ก่อน,นะ,จ้ะ,ใคร,จะ,รอ,เค้า,ดาวน์  
2    เค้า,ดาวน์,ก็,ขอ,ให้,เค้า,ดาน์,อย่าง,มี,ความ,ส...  
3    นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...  
4                          ขอ,ให้,เชิญ,อร่อย,นะ,ทุก,คน  
..                                                 ...  
970  นั้น,แหละ,คุย,กัน,ไป,อูย,คุย,กัน,ไป,เป็น,ปี,อย...  
971  ก็,เลย,แบบ,ชี,วิต,ช่วง,นั้น,แบบ,จะ,เอา,ยัง,ไง,...  
972  เสร็จ,แล้ว,ก็,เลย,อ่า,เขา,เชิญ,ปะ,เออ,เชิญ,มา,...  
973  ช่วง,นั้น,ฉัน,ก็,เคว้ง,อ่ะ,ค่ะ,จะ,ไป,ต่อ,หรือ,...  
974  ฉัน,ก็,เลย,ตัด,สิน,ใจ,กับ,เพื่อน,ว่า,เออ,ฉัน,จ...  

[975 rows x 2 columns]

ข้อสอง

In [240]:
Data.iloc[0][0]

'สวัสดี/สุผู้สุคน/เด้อ/จ้า'

In [241]:
def splitstr(col1):
    co1 = col1.split('/')
    return co1

In [242]:
splitstr(df.iloc[0]['ภาษาอีสาน'])

['ทัก',
 'ทาย',
 'กัน',
 'ซัมบายดี',
 'สวัสดี',
 'วัน',
 'อัง',
 'คาร',
 'เนาะ',
 'จ้า',
 'อา',
 'กาศ',
 'มื้อ',
 'นี่',
 'หนาว',
 'พี',
 'น้อง',
 'แล่ว',
 'ก่อ',
 'อ่า',
 'ช่วง',
 'หน่า',
 'หนาว',
 'นี่',
 'จ้า',
 'ทาง',
 'ทาง',
 'เมือง',
 'นอก',
 'กะ',
 'สิ',
 'แบบ',
 'หว่า']

In [243]:
list_L = []
list_T = []
for i in range(len(Data)):
    if len(splitstr(Data.iloc[i]['ภาษาอีสาน'])) == len(splitstr(Data.iloc[i]['ภาษาไทย'])): # ตัดก่อนแล้วค่อยดู
        list_L += splitstr(Data.iloc[i]['ภาษาอีสาน']) # list_11 = list_11 + splitstr(Data.iloc[i]['ภาษาอีสาน'])
        list_T+= splitstr(Data.iloc[i]['ภาษาไทย'])

In [244]:
len(list_L)

30852

In [245]:
len(list_T)

30852

In [246]:
def Trans(col1,col2):
    str = f'{col1} แปลว่า {col2}' 
    return str

In [247]:
Lang = []
for i in range(len(list_T)):
    if len(list_L[i]) == len(list_T[i]):
        Lang += [lang(list_L[i],list_T[i])] 

In [248]:
T = set(Lang)

In [249]:
T

{'นู แปลว่า นู',
 'ช่วย แปลว่า ช่วย',
 'รับ แปลว่า รับ',
 'ไหน แปลว่า ไหน',
 'เอก แปลว่า เอก',
 'ถ่า แปลว่า ถ้า',
 'จัง แปลว่า ถึง',
 'ดิ่ง แปลว่า ดิ่ง',
 'ตา แปลว่า ตา',
 'ไห แปลว่า ไห',
 'สัพ แปลว่า สัพ',
 'นิ แปลว่า นิ',
 'จ้า แปลว่า จคะ',
 'พูล แปลว่า พูล',
 'แข็ง แปลว่า แข็ง',
 'สปีต แปลว่า สปีต',
 'แว่น แปลว่า แว่น',
 ' มัน แปลว่า  มัน',
 'เส้า แปลว่า เช้า',
 'เด้ แปลว่า จ้า',
 'ใกล้ แปลว่า ใกล้',
 'ล่าง แปลว่า ล้าง',
 'ร่าง แปลว่า ร่าง',
 'เทศกาล แปลว่า เทศกาล',
 'สม่ำ แปลว่า สม่ำ',
 'ทีพ แปลว่า ทีพ',
 'ซาย แปลว่า ชาย',
 'คุม แปลว่า คุม',
 'มืด แปลว่า มืด',
 'เหรียญ แปลว่า เหรียญ',
 'เดี๋ยว แปลว่า เดี่ยว',
 'สุด แปลว่า สุด',
 'ว๊าย แปลว่า ว๊าย',
 'ทาย แปลว่า ทาย',
 'อ่ะ แปลว่า อ่ะ',
 'เฮ็ด แปลว่า เผ็ด',
 'โอ้ แปลว่า โจะ',
 'ทัว แปลว่า ทัว',
 'ลง แปลว่า ลง',
 'แย่ แปลว่า แย่',
 'นี้ แปลว่า นี้',
 'อิ แปลว่า อะ',
 'สิ้น แปลว่า สิ้น',
 'ศุกร์ แปลว่า ศุกร์',
 'ซื่อ แปลว่า ชื้อ',
 'ถาม แปลว่า ถาม',
 'บริษัท แปลว่า บริษัท',
 'แต แปลว่า แต',
 'เกม แปลว่า เกม',
 'ล่ะ แปลว่า ล่ะ',
 'เอ๊า